### ISSUE TITLE
GSConnect crashes Gnome Shell on receiving unicode input, if Caribou is not installed

### ISSUE THREAD

**Issue by jambonmcyeah** [andyholmes/gnome-shell-extension-gsconnect#323]
When I enter a unicode character with remote input on android, gnome-shell crashes.

**Comment by ferdnyc** [comment]
Well, that's not supposed to happen! Sorry to hear that.

Any chance you could reproduce the crash with GSConnect in debug mode, and let us know what's logged when it happens? 

The Wiki has a [preliminary Debug page](https://github.com/andyholmes/gnome-shell-extension-gsconnect/wiki/Debugging) which isn't very complete, but basically:
1. Open the GSConnect Mobile Settings window
2. Select "Debugger" from the hamburger menu at the top right
3. Hit the circular button _next to_ the On/Off switch

That'll launch a terminal window showing the journal for the extension, in debug mode. Then if you can attempt to reproduce the issue again, even though Gnome Shell crashes the logs should hopefully have information about the error.

**Comment by jambonmcyeah** [comment]
Here's the log
```
DEBUG: [lan.js:_onIncomingIdentity:222]: already connected
DEBUG: [lan.js:_onIncomingIdentity:198]: {
  "id": 1541972271793,
  "type": "kdeconnect.identity",
  "body": {
    "deviceId": "9897835f5e3acfb3",
    "deviceName": "Xiaomi Redmi Note 4",
    "protocolVersion": 7,
    "deviceType": "phone",
    "incomingCapabilities": [
      "kdeconnect.sms.request_conversations",
      "kdeconnect.telephony.request_mute",
      "kdeconnect.telephony.request",
      "kdeconnect.mpris",
      "kdeconnect.notification",
      "kdeconnect.sms.request_conversation",
      "kdeconnect.findmyphone.request",
      "kdeconnect.ping",
      "kdeconnect.systemvolume",
      "kdeconnect.notification.reply",
      "kdeconnect.share.request",
      "kdeconnect.sftp.request",
      "kdeconnect.notification.request",
      "kdeconnect.mousepad.request",
      "kdeconnect.contacts.request_vcards_by_uid",
      "kdeconnect.sms.request",
      "kdeconnect.runcommand",
      "kdeconnect.battery.request",
      "kdeconnect.clipboard",
      "kdeconnect.contacts.request_all_uids_timestamps"
    ],
    "outgoingCapabilities": [
      "kdeconnect.sms.messages",
      "kdeconnect.telephony",
      "kdeconnect.notification",
      "kdeconnect.contacts.response_uids_timestamps",
      "kdeconnect.findmyphone.request",
      "kdeconnect.ping",
      "kdeconnect.mousepad.keyboardstate",
      "kdeconnect.share.request",
      "kdeconnect.contacts.response_vcards",
      "kdeconnect.notification.request",
      "kdeconnect.mousepad.echo",
      "kdeconnect.mousepad.request",
      "kdeconnect.sftp",
      "kdeconnect.runcommand.request",
      "kdeconnect.mpris.request",
      "kdeconnect.systemvolume.request",
      "kdeconnect.battery",
      "kdeconnect.clipboard"
    ],
    "tcpPort": 1716,
    "tcpHost": "192.168.2.14"
  }
}
DEBUG: [lan.js:_onIncomingIdentity:222]: already connected
DEBUG: [core.js:receive/<:404]: Xiaomi Redmi Note 4: {
  "id": 1541972277671,
  "type": "kdeconnect.mousepad.request",
  "body": {
    "key": "€"
  }
}
```

**Comment by ferdnyc** [comment]
And then it crashes right after that? Hmm.

Can you also check that the "Extended Keyboard Support" dependencies are installed for your OS? If you check the "Other" panel of the main Mobile Settings window, it's on the list of "Additional Features". It'll either show a check mark, if everything's installed, or hopefully an install button if there are dependencies it needs. If there are any missing, that might explain the crash.

(BTW, what OS/distro are you running on?)

**Comment by jambonmcyeah** [comment]
I didn't have caribou. So I installed it and it works. My suggestion is give a warning if caribou isn't installed instead of crashing.
I'm using Arch btw.

**Comment by ferdnyc** [comment]
Oh, no, absolutely it's not supposed to _crash_ if it's missing! Thanks for reporting that, and glad to hear it was that simple.

**Comment by andyholmes** [comment]
Hmm, maybe a better way is to detect Unicode characters and just bail entirely. Aside from being a deprecated library, that is probably unmaintained, Caribou is an X11/Xorg only thing that we'll probably have to drop eventually anyways.

@jambonmcyeah Is there an actual stack trace in your logs somewhere about the crash itself? A bad call to AT-SPI shouldn't be able to take down the desktop, so even if we have to work around it for now an issue should be opened upstream to report this.

**Comment by andyholmes** [comment]
This seems a little trickier than I thought. I can confirm this will kill an X session, just working through this with the GJS maintainer; no need to get me another log.

**Comment by seb128** [comment]
Did you ever upstream that at-spi or Xorg? If the Xserver is segfaulting it seems a problem on their side. 
https://launchpad.net/bugs/1809737 is a somewhat similar issue with onboard, which was upstreamed now to https://gitlab.freedesktop.org/xorg/xserver/issues/679 (just for information in case you were still interested by getting the issue properly resolved)

**Comment by andyholmes** [comment]
TL;DR it's probably a combination of at-spi2-core, gnome-shell and Xorg, rather than just one of those. Probably gnome-shell is remapping the XKeyMap while at-spi2-core is, resulting in some unmapped key event being sent.

The cause of this is not entirely clear. libfakekey works for this type of input, however it does it's own on-the-fly keymap changes, so I suspect this is actually a problem between at-spi2 and gnome-shell (which also does it's own key mapping I think). This is in fact why we dropped support for libcaribou, which is essentially a wrapper around XTest (used by at-spi2 and libfakekey) and has similar problems.

Most likely this could be fixed in at-spi2-core by doing some custom re-mapping to work around whatever gnome-shell does, however it seems like AT-SPI has been in "maintenance mode" for some time and even libcaribou is deprecated. The current iteration of gnome-shell actually built key synthesis into the version of clutter it ships with and that's what the new OSK uses. There's been very little talk about what to do about accessibility/at-spi2-core in Wayland given the security models and so on, but in general AT-SPI seems to be pretty stagnant.
